controllers/class.reactcontroller.php

```diff
@@ -49,25 +49,23 @@ public function Index($Type, $ID, $ActionID) {
       throw PermissionException();
     }
 
-    switch($Type) {
-      case 'discussion':
-        $Model = new DiscussionModel();
-        $AnchorID = '#Discussion_';
-        break;
-      case 'comment':
-        $Model = new CommentModel();
-        $AnchorID = '#Comment_';
-        break;
-      case 'activity':
-        $Model = new ActivityModel();
-        $AnchorID = '#Activity_';
-        break;
-      default:
-        throw new Gdn_UserException(T('Yaga.Action.InvalidTargetType'));
-        break;
-    }
+    $Item = null;
+    $AnchorID = '#' . ucfirst($Type) . '_';
+    $ItemOwnerID = 0;
 
-    $Item = $Model->GetID($ID);
+    if(in_array($Type, array('discussion', 'comment', 'activity'))) {
+      $Item = GetRecord($Type, $ID);
+    }
+    else {
+      $this->EventArguments = array(
+        'TargetType' => $Type,
+        'TargetID' => $ID,
+        'Item' => &$Item,
+        'AnchorID' => &$AnchorID,
+        'ItemOwnerID' => &$ItemOwnerID
+      );
+      $this->FireEvent('UnknownType');
+    }
```

Comment:
I would prefer to call the even `CustomType` so the hook is `reactController_customType_handler`.

More importantly there should be a `TypeFound` event argument that plugins should set to true. Then check it right after the event and throw an exception if the type wasn't found.